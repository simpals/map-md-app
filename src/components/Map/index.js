import React, { useEffect, useState, Fragment } from 'react';
import { createPortal } from 'react-dom';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import mapboxgl from 'mapbox-gl';
import objects3DRender from '../../Render3D';
import MapControllers from '../MapControllers';
import Logo from '../Logo';
import { MapCtx } from '../contexts';

import 'mapbox-gl/dist/mapbox-gl.css';
import '../../static/css/index.css';

const MapContainer = styled.div`
  width: 100%;
  height: 100%;
`;

const setAuthHeader = (url, authorization) => {
  if (url.indexOf('google') > -1) {
    return { url };
  }
  return {
    url,
    headers: {
      Authorization: authorization
    }
  };
};

const Map = (props) => {
  const {
    mapStyles,
    authorization,
    viewport,
    onInit,
    onUserLocInit,
    portalNode,
    init3DObjects,
    disable3DMap,
    children,
    mapImages,
    utm,
    ...rest
  } = props;

  const {
    pitch,
    bearing,
    zoom,
    minZoom,
    maxZoom,
    center,
    maxBounds: [[S, W], [N, E]],
    initialStyle
  } = viewport;

  const [MapGL, setMapGL] = useState(null);
  const [userLocation, setUserLocationApi] = useState(null);
  const [currentStyle, setStyle] = useState(initialStyle);
  const changeCurentStyle = newStyle => setStyle(newStyle);

  const mapOnLoad = (map) => {
    map.once('load', () => {
      onInit(map);
    });
  };

  const initUserLocation = (mapCtx) => {
    const userLocationAPI = new mapboxgl.GeolocateControl({
      positionOptions: {
        enableHighAccuracy: true
      },
      trackUserLocation: true
    });
    mapCtx.addControl(userLocationAPI, 'top-left');
    setUserLocationApi(userLocationAPI);
    onUserLocInit(userLocationAPI);
  };

  const initMap = () => {
    const southWest = new mapboxgl.LngLat(S, W);
    const northEast = new mapboxgl.LngLat(N, E);
    const bounds = new mapboxgl.LngLatBounds(southWest, northEast);
    const map = new mapboxgl.Map({
      container: 'map-app-container',
      style: mapStyles,
      center,
      maxBounds: bounds,
      zoom,
      minZoom,
      maxZoom,
      pitch,
      bearing,
      fadeDuration: 0,
      transformRequest: url => setAuthHeader(url, authorization)
    });
    map.touchZoomRotate.enable();
    initUserLocation(map);
    mapOnLoad(map);
    setMapGL(map);
    mapImages.map((image) => {
      const { id, path } = image;
      map.loadImage(path, (_, imageFile) => {
        if (!map.hasImage(id) && imageFile) map.addImage(id, imageFile);
      });
    });
    if (init3DObjects && !disable3DMap) objects3DRender(map);
  };

  useEffect(() => {
    initMap();
  }, []);

  return (
    <Fragment>
      <MapContainer id='map-app-container' />
      <MapCtx.Provider
        value={{
          MapGL,
          userLocation,
          authorization,
          style: currentStyle,
          changeStyle: changeCurentStyle
        }}
      >
        <Logo {...utm} />
        {MapGL && userLocation && (
          <MapControllers {...rest} disable3DMap={disable3DMap} style={initialStyle} mapImages={mapImages} />
        )}
        {MapGL && portalNode && !!children && createPortal(children({ MapGL }), portalNode)}
      </MapCtx.Provider>
    </Fragment>
  );
};

Map.defaultProps = {
  onInit: () => {},
  onUserLocInit: () => {},
  portalNode: null,
  children: null,
  init3DObjects: false,
  disable3DMap: false,
  mapImages: [],
  utm: {}
};

Map.propTypes = {
  authorization: PropTypes.string.isRequired,
  mapStyles: PropTypes.object.isRequired,
  viewport: PropTypes.object.isRequired,
  onInit: PropTypes.func,
  onUserLocInit: PropTypes.func,
  portalNode: PropTypes.instanceOf(React.Element),
  children: PropTypes.func,
  init3DObjects: PropTypes.bool,
  disable3DMap: PropTypes.bool,
  mapImages: PropTypes.array,
  utm: PropTypes.object
};

export default Map;
