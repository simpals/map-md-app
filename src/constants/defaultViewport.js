export const DEFAULT_VIEWPORT = {
  pitch: 0,
  bearing: 0,
  zoom: 10,
  minZoom: 7,
  maxZoom: 20,
  center: [28.83917, 47.0173],
  maxBounds: [[24.0, 45.0], [32.0, 49.0]],
  initialStyle: 'map'
};
