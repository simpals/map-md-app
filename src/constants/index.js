export * from './optionsStyle';
export * from './defaultViewport';
export * from './options3D';
