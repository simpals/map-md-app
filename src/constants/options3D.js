export const URL_3D = 'https://map.md/api/api/object/list';

export const URL_3D_FILE = 'https://map.md/api/images/map/objects/';
