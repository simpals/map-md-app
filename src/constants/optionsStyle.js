export const MAP_STYLE = {
  URL: {
    map: 'https://map.md/api/tiles/styles/map/style.json',
    satelite: 'https://map.md/api/tiles/styles/satelite/style.json',
  },
  OPTIONS: {
    mode: 'cors',
    headers: {
      authorization: null
    }
  }
};
