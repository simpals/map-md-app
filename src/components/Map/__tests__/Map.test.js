import 'jsdom-global/register';
import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { DEFAULT_VIEWPORT } from '../../../constants';

import Map from '..';

jest.mock('mapbox-gl', () => ({
  GeolocateControl: jest.fn(),
  Map: jest.fn(() => ({
    addControl: jest.fn(),
    on: jest.fn(),
    once: jest.fn(),
    getPitch: jest.fn(),
    touchZoomRotate: {
      enable: jest.fn()
    }
  })),
  LngLat: jest.fn(),
  LngLatBounds: jest.fn(),
}));

const setApiKey = ''; // API key
const ENCODE = Buffer.from(`${setApiKey}:pashalka`).toString('base64');
const AUTHORIZATION = `Basic ${ENCODE}`;

const props = {
  authorization: AUTHORIZATION,
  mapStyles: {},
  viewport: DEFAULT_VIEWPORT
};

describe('test Map component', () => {
  it('renders without crashing', () => {
    const component = shallow(<Map {...props} />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('test Map component useEffect', (done) => {
    const component = mount(<Map {...props} />);
    setImmediate(() => {
      component.update();
      done();
    });
    expect(component.childAt(2).name()).toEqual('MapControllers');
  });

  it('test Map component context', (done) => {
    const component = shallow(<Map {...props} />);
    const context = component.children().find({ value: {} });
    if (!context.debug()) done.fail(new Error('Context not found'));
    done();
  });
});
