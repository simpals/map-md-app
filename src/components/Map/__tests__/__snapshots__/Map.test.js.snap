// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Map component renders without crashing 1`] = `
<Fragment>
  <styled.div
    id="map-app-container"
  />
  <ContextProvider
    value={
      Object {
        "MapGL": null,
        "authorization": "Basic OGMwZTY5M2MtNTU1Yi00ZGJmLThlMTAtYWM4NjQ4YThjODU0OnBhc2hhbGth",
        "changeStyle": [Function],
        "style": "map",
        "userLocation": null,
      }
    }
  >
    <Logo
      campaign=""
      content=""
      medium=""
      source=""
    />
  </ContextProvider>
</Fragment>
`;
